### PR TITLE
KD trainer w/ logprobs

### PR DESCRIPTION
**Pull request by winglian**
# Description

### CODE REVIEW

**Comment by winglian** [comment]
see https://wandb.ai/axolotl-ai/logprob-kd/runs/uj32mbv1/files/tmp/axolotl_config_301w_dfz.yml

File: src/axolotl/integrations/kd/trainer.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        num_items_in_batch=None,
+        shift_targets=False,
+    ):
+        """
+        How the loss is computed by Trainer. By default, all models return the loss in the first element.
+
+        Subclass and override for custom behavior.
+        """
+
+        target_logprobs = inputs.pop("target_logprobs")
```
Comment (salmanmohammadi) [inline src/axolotl/integrations/kd/trainer.py:56]: Could you ELI5 the choice of using logprobs vs logits for the teacher?

Comment (salmanmohammadi) [reply src/axolotl/integrations/kd/trainer.py:56]: Is this for offline KD?

Comment (winglian) [reply src/axolotl/integrations/kd/trainer.py:56]: > Could you ELI5 the choice of using logprobs vs logits for the teacher?

It's quite easy for users to get logprobs, since that's provided out of the box by vllm. No custom inference solution required, whereas logits usually requires a little bit of custom work.

Comment (winglian) [reply src/axolotl/integrations/kd/trainer.py:56]: > Is this for offline KD?

Yeah. Offline KD of the teacher logprobs.

File: src/axolotl/core/trainers/base.py [hunk 2]
```diff
@@ -0,0 +1,1034 @@
...
+    def _save_checkpoint(self, model, trial, **kwargs):
+        # make sure the checkpoint dir exists, since trainer is flakey
+        checkpoint_folder = f"{PREFIX_CHECKPOINT_DIR}-{self.state.global_step}"
+        run_dir = self._get_output_dir(trial=trial)
+        output_dir = os.path.join(run_dir, checkpoint_folder)
+        os.makedirs(output_dir, exist_ok=True)
+        return super()._save_checkpoint(model, trial, **kwargs)
+
+
+class AxolotlMambaTrainer(AxolotlTrainer):
```
Comment (salmanmohammadi) [inline src/axolotl/core/trainers/base.py:736]: Not for this PR, but how would we feel about moving each of these trainers to their own file?

Comment (salmanmohammadi) [reply src/axolotl/core/trainers/base.py:736]: having a `integrations/trl/` folder would be neat too.

File: src/axolotl/cli/args.py [hunk 3]
```diff
@@ -13,6 +13,12 @@ class PreprocessCliArgs:
     debug_num_examples: int = field(default=1)
     prompter: Optional[str] = field(default=None)
     download: Optional[bool] = field(default=True)
+    iterable: Optional[bool] = field(
```
Comment (salmanmohammadi) [inline src/axolotl/cli/args.py:16]: Is it worth documenting somewhere (or raising an issue as a reminder) to show users that we offer support for iterable datasets?

File: src/axolotl/cli/main.py [hunk 4]
```diff
@@ -39,6 +39,8 @@ def preprocess(config: str, **kwargs) -> None:
         kwargs: Additional keyword arguments which correspond to CLI args or `axolotl`
             config options.
     """
+    kwargs = {k: v for k, v in kwargs.items() if v is not None}
```
Comment (salmanmohammadi) [inline src/axolotl/cli/main.py:42]: Shouldn't using `@filter_none_kwargs` address this?

File: src/axolotl/integrations/kd/topk_logprob/LICENSE.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
+### AXOLOTL COMMUNITY LICENSE AGREEMENT
```
Comment (salmanmohammadi) [inline src/axolotl/integrations/kd/topk_logprob/LICENSE.md:1]: Is it worth surfacing this community license somewhere in our docs/README and how it would be used? Fine to leave as a follow up/issue.

File: src/axolotl/integrations/kd/args.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+from pydantic import BaseModel
+
+
+class KDArgs(BaseModel):
+    """
+    Input args for knowledge distillation.
+    """
+
+    kd_trainer: Optional[bool] = None  # whether to use KD trainer
```
Comment (salmanmohammadi) [inline src/axolotl/integrations/kd/args.py:28]: Slightly confused - what happens when `kd_trainer=False`?

File: src/axolotl/utils/trainer.py [hunk 7]
```diff
@@ -95,9 +95,46 @@ def disable_datasets_caching():
...
+    if "input_ids" not in sample:
+        # If there's no "input_ids", just return sample unchanged
+        return sample
+
+    input_ids = sample["input_ids"]
+
+    # Detect if it's a single example or a batch
+    if not input_ids:
+        # Edge case: empty
+        return sample
```
Comment (salmanmohammadi) [inline src/axolotl/utils/trainer.py:112]: nit
```suggestion
    # Return sample unchanged if "input_ids" is not present, or is empty
    if "input_ids" not in sample or not sample["input_ids"]:
        return sample

    input_ids = sample["input_ids"]
```

File: src/axolotl/utils/trainer.py [hunk 8]
```diff
@@ -172,10 +209,31 @@ def add_length(sample):
...
+    """
+    Drop samples whose sequence length is either too long (> sequence_len)
+    or too short (< min_sequence_len).
+
+    Works for both single-example (list[int]) or batched (list[list[int]]).
+    """
+    input_ids = sample["input_ids"]
+
+    # Edge case: if input_ids is empty
+    if not input_ids:
```
Comment (salmanmohammadi) [inline src/axolotl/utils/trainer.py:221]: Should you do the same check above for `not sample["input_ids"]`?

File: src/axolotl/utils/trainer.py [hunk 9]
```diff
@@ -172,10 +209,31 @@ def add_length(sample):
 
 
 def drop_long_seq(sample, sequence_len=2048, min_sequence_len=2):
```
Comment (salmanmohammadi) [inline src/axolotl/utils/trainer.py:211]: nit: this is more of a fiiltering function with the signature `def filter_sequence_length(...) -> Union[bool, List[bool]]` right?

File: src/axolotl/utils/trainer.py [hunk 10]
```diff
@@ -185,10 +243,13 @@ def process_datasets_for_packing(cfg, train_dataset, eval_dataset):
...
-    LOG.debug(f"min_input_len: {min_input_len}", main_process_only=True)
-    max_input_len = np.max(get_dataset_lengths(train_dataset))
-    LOG.debug(f"max_input_len: {max_input_len}", main_process_only=True)
+    try:
+        min_input_len = np.min(get_dataset_lengths(train_dataset))
+        LOG.debug(f"min_input_len: {min_input_len}", main_process_only=True)
+        max_input_len = np.max(get_dataset_lengths(train_dataset))
+        LOG.debug(f"max_input_len: {max_input_len}", main_process_only=True)
+    except AttributeError:
+        pass
```
Comment (salmanmohammadi) [inline src/axolotl/utils/trainer.py:252]: Is there anything informative worth logging here?

File: src/axolotl/utils/trainer.py [hunk 11]
```diff
@@ -203,59 +264,108 @@ def process_datasets_for_packing(cfg, train_dataset, eval_dataset):
...
-    prior_len = len(train_dataset)
+    filter_map_kwargs = {}
+    if not isinstance(train_dataset, IterableDataset):
+        filter_map_kwargs["num_proc"] = cfg.dataset_processes
+        filter_map_kwargs["load_from_cache_file"] = not cfg.is_preprocess
+
+    try:
+        prior_len = len(train_dataset)
+    except TypeError:
+        # handle iterable datasets case
```
Comment (salmanmohammadi) [inline src/axolotl/utils/trainer.py:275]: How come an `isinstance` check like above wouldn't work here?

File: src/axolotl/utils/trainer.py [hunk 12]
```diff
@@ -203,59 +264,108 @@ def process_datasets_for_packing(cfg, train_dataset, eval_dataset):
...
+        labels = sample["labels"]
+        if not labels:
+            return True
+
+        # Check if single example or batch
+        # If first element is an int, we assume a single example
+        # If it's a list, we assume we're dealing with a batch
+        if isinstance(labels[0], int):
+            # Single example: return a single bool
+            return np.sum(np.array(labels) != -100) > 0
```
Comment (salmanmohammadi) [inline src/axolotl/utils/trainer.py:321]: nit 
```suggestion
            return np.any(labels != -100)
```

File: src/axolotl/utils/trainer.py [hunk 13]
```diff
@@ -203,59 +264,108 @@ def process_datasets_for_packing(cfg, train_dataset, eval_dataset):
...
+
+        # Batched: 'labels' is a list of lists
+        # Return a list of booleans, one per sub-list
+        results = []
+        for row_labels in labels:
+            # Each row_labels is a list[int]
+            results.append(np.sum(np.array(row_labels) != -100) > 0)
```
Comment (salmanmohammadi) [inline src/axolotl/utils/trainer.py:328]: nit
```suggestion
        results = [np.any(row_labels != -100) for row_labels in labels]
```

File: tests/e2e/integrations/test_kd.py [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+        "torch_compile": True,
+        "chat_template": "llama3",
+        "kd_trainer": True,
+        "kd_ce_alpha": 0.1,
+        "kd_alpha": 0.9,
+        "kd_temperature": 2.0,
+        "dataloader_prefetch_factor": 8,
+        "dataloader_num_workers": 4,
+        "dataloader_pin_memory": True,
+        # "dataset_prepared_path": str(Path(temp_dir) / "last_run_prepared"),
```
Comment (salmanmohammadi) [inline tests/e2e/integrations/test_kd.py:36]:
```suggestion
```
might as well chop while we're here

File: src/axolotl/utils/data/shared.py [hunk 15]
```diff
@@ -29,7 +29,9 @@ def get_ds_type(config_dataset: DictDefault):
     return ds_type
 
 
-def load_dataset_w_config(config_dataset, auth_token):
+def load_dataset_w_config(
+    config_dataset, auth_token, streaming=False
```
Comment (salmanmohammadi) [inline src/axolotl/utils/data/shared.py:33]: Worth folding `streaming` into `config_dataset` so you're doing `config_dataset.streaming`?

File: src/axolotl/integrations/kd/topk_logprob/forward_kl.py [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+    A KD loss function that is TorchScript-friendly.
+    """
+
+    # Determine the teacher sequence length
+    # target_token_ids shape: [B, teacher_seq_len, K]
+    # student_logits shape:   [B, student_seq_len, vocab_size]
+    teacher_seq_len = target_token_ids.shape[1]
+
+    # Slice student logits to match teacher-provided sequence length
+    student_logits_for_kd = student_logits[
```
Comment (salmanmohammadi) [inline src/axolotl/integrations/kd/topk_logprob/forward_kl.py:87]: When would the student be predicting a longer sequence length?

**Comment by salmanmohammadi** [comment]
I've taken a first pass and it looks pretty good overall. I think the KD logic checks out. I'll make another pass tomorrow.

File: src/axolotl/integrations/kd/topk_logprob/forward_kl.py [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+def loss(
+    student_logits: torch.Tensor,
+    target_token_ids: torch.Tensor,
+    target_logprobs: torch.Tensor,
+    target_mask: torch.Tensor,
+    num_items_in_batch: int = -1,  # Use -1 to indicate "None"
+    kd_temperature: float = 1.0,
+) -> torch.Tensor:
+    """
+    A KD loss function that is TorchScript-friendly.
```
Comment (salmanmohammadi) [inline src/axolotl/integrations/kd/topk_logprob/forward_kl.py:63]: Could you add docs for what the parameters are? I'd mainly like to clarify what `target_mask` is used for but might as well doc them all : )

File: src/axolotl/cli/main.py (hunk 4 above)
Comment (winglian) [reply src/axolotl/cli/main.py:42]: yeah, probably an artifact of a rebase

File: src/axolotl/integrations/kd/args.py (hunk 6 above)
Comment (winglian) [reply src/axolotl/integrations/kd/args.py:28]: same as no kd, so should default to sft. happy to workshop a better way to enable kd

File: src/axolotl/utils/trainer.py (hunk 9 above)
Comment (winglian) [reply src/axolotl/utils/trainer.py:211]: correct, it used to only be for samples with a sequence length longer than `sequence_len` and evolved to drop things shorter than the minimum.

File: src/axolotl/utils/trainer.py (hunk 10 above)
Comment (winglian) [reply src/axolotl/utils/trainer.py:252]: not really, the code within the try block is really mostly debugging about the lengths of the inputs, and iirc fails in the context of an iterable dataset.

File: src/axolotl/utils/data/shared.py (hunk 15 above)
Comment (winglian) [reply src/axolotl/utils/data/shared.py:33]: streaming comes from the cli `iterable`, so not a dataset configuration arg

File: src/axolotl/integrations/kd/topk_logprob/forward_kl.py (hunk 16 above)
Comment (winglian) [reply src/axolotl/integrations/kd/topk_logprob/forward_kl.py:87]: allows for logprob datasets that only provide logprobs on the response/assistant outputs and didn't capture logprobs for the inputs.
